be/src/storage/primary_key_encoder.cpp

```diff
@@ -455,13 +458,39 @@ void PrimaryKeyEncoder::encode_sort_key(const VectorizedSchema& schema, const Ch
     prepare_ops_datas(schema, schema.sort_key_idxes(), chunk, &ops, &datas);
     auto& bdest = down_cast<BinaryColumn&>(*dest);
     bdest.reserve(bdest.size() + len);
+    std::vector<std::shared_ptr<Column>> cols(ncol);
+    for (int i = 0; i < ncol; i++) {
+        cols[i] = chunk.get_column_by_index(schema.sort_key_idxes()[i]);
+    }
+    bool has_nullable_sort_key = false;
+    for (int i = 0; i < ncol; i++) {
+        if (schema.field(schema.sort_key_idxes()[i])->is_nullable()) {
+            has_nullable_sort_key = true;
+            break;
+        }
+    }
     std::string buff;
-    for (size_t i = 0; i < len; i++) {
-        buff.clear();
-        for (int j = 0; j < ncol; j++) {
-            ops[j](datas[j], offset + i, &buff);
+    if (!has_nullable_sort_key) {
+        for (size_t i = 0; i < len; i++) {
+            buff.clear();
+            for (int j = 0; j < ncol; j++) {
+                ops[j](datas[j], offset + i, &buff);
+            }
+            bdest.append(buff);
+        }
+    } else {
+        for (size_t i = 0; i < len; i++) {
+            buff.clear();
+            for (int j = 0; j < ncol; j++) {
+                if (cols[j]->is_null(i)) {
+                    buff.push_back(SORT_KEY_NULL_FIRST_MARKER);
```

Comment:
If the data is also started as 0x00, how to handle it?

Reply:
> If the data is also started as 0x00, how to handle it?

If we're a middle component of a composite key, we need to add a \x00 at the end in order to separate this component from the next one. However, if we just did that, we'd have issues where a key that actually has \x00 in it would compare wrong, so we have to instead add \x00\x00, and encode \x00 as \x00\x01.

Reply:
Yes. So just add \x00 is not right.